llama_stack/apis/tools/tools.py

```diff
@@ -0,0 +1,117 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the terms described in the LICENSE file in
+# the root directory of this source tree.
+
+from typing import Annotated, Any, Dict, List, Literal, Optional, Union
+
+from llama_models.llama3.api.datatypes import ToolPromptFormat
+from llama_models.schema_utils import json_schema_type, register_schema, webmethod
+from pydantic import BaseModel, Field
+from typing_extensions import Protocol, runtime_checkable
+
+from llama_stack.apis.inference import InterleavedContent
+
+from llama_stack.apis.resource import Resource, ResourceType
+from llama_stack.providers.utils.telemetry.trace_protocol import trace_protocol
+
+
+@json_schema_type
+class ToolParameter(BaseModel):
+    name: str
+    parameter_type: str
+    description: str
+
+
+@json_schema_type
+class Tool(Resource):
+    type: Literal[ResourceType.tool.value] = ResourceType.tool.value
+    name: str
+    tool_group: str
+    description: str
+    parameters: List[ToolParameter]
+    provider_id: Optional[str] = None
+    tool_prompt_format: Optional[ToolPromptFormat] = Field(
+        default=ToolPromptFormat.json
+    )
+
+
+@json_schema_type
+class ToolDef(BaseModel):
+    name: str
+    description: str
+    parameters: List[ToolParameter]
+    metadata: Dict[str, Any]
+    tool_prompt_format: Optional[ToolPromptFormat] = Field(
+        default=ToolPromptFormat.json
+    )
+
+
+@json_schema_type
+class MCPToolGroup(BaseModel):
+    type: Literal["mcp"] = "mcp"
+    endpoint: str
+
+
+@json_schema_type
+class UserDefinedToolGroup(BaseModel):
+    type: Literal["user_defined"] = "user_defined"
+    tools: List[ToolDef]
+
+
+ToolGroup = register_schema(
+    Annotated[Union[MCPToolGroup, UserDefinedToolGroup], Field(discriminator="type")],
+    name="ToolGroup",
+)
+
+
+@json_schema_type
+class InvokeToolResult(BaseModel):
+    content: InterleavedContent
+    error_message: Optional[str] = None
+    error_code: Optional[int] = None
+
+
+class ToolStore(Protocol):
+    def get_tool(self, tool_id: str) -> Tool: ...
+
+
+@runtime_checkable
+@trace_protocol
+class Tools(Protocol):
```

Comment:
should be `ToolGroups`

Reply:
only the register function takes in a tool group. the other api operates on tool itself. Even in the registry, its the tool that gets registered. I am not sure if we should name this API as `ToolGroup`.

Reply:
I think this should definitely be `ToolGroups` since that's the entity the client will register / unregister and list.